content/articles/federated-learning-using-tensorflow-federated/index.md

```diff
@@ -0,0 +1,201 @@
+            loss=tf.keras.losses.SparseCategoricalCrossentropy(),
+            optimizer=tf.keras.optimizers.SGD(learning_rate=0.02),
+            metrics=[tf.keras.metrics.SparseCategoricalAccuracy()] 
+
+        )
+    return model
+```
+
+If you're familiar with TensorFlow, the above Keras code already looks familiar to you. We will import it into our main function `model_fn` using the `create_compiled_keras_model()` method.
+
+```python
+    def model_fn():
+
+        keras_model = create_compiled_keras_model()
+
+        return tff.learning.from_compiled_keras_model(keras_model, sample_batch)
+```
+The above code represents where you will plug in your own Keras model.
+
+```python
+     state = train.initialize() 
+
+    for _ in range (5):
+        state, metrics = train.next(state, train_data)
+        print (metrics.loss) 
+
+```
+
+In the above code, the `initialize()` method will get our initial server state. It then calls `train.next` which will run our federated training. So, this includes sending the initial server state to each of the clients. Each of the clients will run their own local rounds of training and send their updates back to the server. Back on the server, we'll get this new aggregated global model that's been produced from the decentralized data of each of the clients.  
+
+```python
+    eval = tff.learning.build_federated_evaluation(model_fn)
+    metrics = eval(state.model, test_data)
+```
+
+Finally, we can perform federated evaluation to understand the state of our trained model. The method `build_federated_evaluation()` method will provide federated evaluation for you. 
+
+Here's how the whole code looks like for TFF:
+
+```python
+
+    train_data, test_data = 
+    tff.simulation.datasets.emnist.load_data()
+
+    def model_fn():
+
+        keras_model = create_compiled_keras_model()
+
+        return tff.learning.from_compiled_keras_model(keras_model, sample_batch)
+    
+    train = tff.learning.build_federated_averaging_process(model_fn)
+
+    state = train.initialize() 
+
+    for _ in range (5):
+        state, metrics = train.next(state, train_data)
+        print (metrics.loss) 
+
+    eval = tff.learning.build_federated_evaluation(model_fn)
+    metrics = eval(state.model, test_data)
+```
+In summary, the general components for the FL API include:
+
+1. Models
+
+- `tff.learning.Model`
+
+- `create_compiled_keras_model()`
+
+2. Federated computation builders
+
+TFF provides two builder functions:
+
+- `tff.learning.build_federated_averaging_process` generates the federated computations for federated training.
+
+- `tff.learning.build_federated_evaluation` generates the federated computations for federated evaluation.
+
+To avoid making the article too long, I have implemented Federated Learning for Image classification using Tensorflow Federated on my [Google Colab](https://colab.research.google.com/drive/1EuSVn6gVVKy0pI_m1nooCd181v6ym5J-?authuser=1#scrollTo=0QIjHnqpGUy4). Please check it out.
```

Comment:
I believe it would be of great value adding some more step-by-step content to bring this tutorial closer to your Colab. i.e.

Include a step taking the reader through the dataset being used.
What simulation dataset is being used?

On your colab you mention using a subset of client devices. Take the reader through how this happens. Hold a reader's hand through training.

You mention that it's FL for image classification. That is the context the reader should get that right off the bat.

Finally, the output is:
_12.931682
13.18153
12.487086
12.092873
11.236066_
A reader is likely to wonder how they are to interpret these results. Do these metrics imply success? Has FL happened?